### ISSUE TITLE
Java Grammar: multiple variable declaration tree doesn't look natural

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3151]
````
$ cat TestClass.java
public class TestClass {
    int var6 = 5; int var7 = 6, //indent:4 exp:4
        var8 = 5; //indent:8 exp:8
}


$ java -jar checkstyle-6.18-all.jar -t TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
    |  |  `--LITERAL_INT -> int [2:4]
    |  |--IDENT -> var6 [2:8]
    |  |--ASSIGN -> = [2:13]
    |  |  `--EXPR -> EXPR [2:15]
    |  |      `--NUM_INT -> 5 [2:15]
    |  `--SEMI -> ; [2:16]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:18]
    |  |--MODIFIERS -> MODIFIERS [2:18]
    |  |--TYPE -> TYPE [2:18]
    |  |  `--LITERAL_INT -> int [2:18]
    |  |--IDENT -> var7 [2:22]
    |  |--ASSIGN -> = [2:27]
    |  |  `--EXPR -> EXPR [2:29]
    |  |      `--NUM_INT -> 6 [2:29]
    |  `--SEMI -> ; [3:16]
    |--COMMA -> , [2:30]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:18]
    |  |--MODIFIERS -> MODIFIERS [2:18]
    |  |--TYPE -> TYPE [2:18]
    |  |  `--LITERAL_INT -> int [2:18]
    |  |--SINGLE_LINE_COMMENT -> // [2:32]
    |  |  `--COMMENT_CONTENT -> indent:4 exp:4\n [2:34]
    |  |--IDENT -> var8 [3:8]
    |  `--ASSIGN -> = [3:13]
    |      `--EXPR -> EXPR [3:15]
    |          `--NUM_INT -> 5 [3:15]
    |--SINGLE_LINE_COMMENT -> // [3:18]
    |  `--COMMENT_CONTENT -> indent:8 exp:8\n [3:20]
    `--RCURLY -> } [4:0]
````
I was expecting variables 7 and 8 to either have the same parent, or have some continuity with the order and where the semi-colon would be. As it is now, it looks like `var7` has the semi-colon and `var8` ends with nothing. This also causes the line numbers to look out of order, as `var7` ends with line 3, we have a comma at line 2, and then `var8` starts at line 2.

**Comment by ajayramesh23** [comment]
I am new this eco-system, so picking easy ones related to _antlr_. 
@rnveach  - Can I work on this issue ?

**Comment by rnveach** [comment]
@ajayramesh23 This one was not tagged as `easy`.
Any changes to ANTLR will have drastic repercussions to all our checks, including external 3rd party checks.
I recommend not touching any ANTLR until you have done a few fixes to get at ease with our review process.

@romani I recommend with mark this with checkstyle 8 label as it will break backwards compatibility.
I will examine other antlr issues.

**Comment by romani** [comment]
All antlr issues are hard issues, as they required a lot of testing and might cause a lot of changes on Checks or ........... . If user is new to project I would recommend to fix smth in Checks first to get to know how all that tokens are used and what is critical in processing of them.

**Comment by romani** [comment]
@nmancus1 , please give us insight on what we can do to address this.

**Comment by nmancus1** [comment]
From https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L825-L829

If we match the SEMI before building `variableDefinitions`, then append add it as a child, like this:
``` |    (s6:SEMI)? v:variableDefinitions[#mods,#t]
                           {
                               #field = #v;
                               #v.addChild(#s6);
                           }
```

It will produce a tree like this:

```
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
...
    |--SEMI -> ; [2:16]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:18]
    |   |--MODIFIERS -> MODIFIERS [2:18]
    |   |--TYPE -> TYPE [2:18]
    |   |   `--LITERAL_INT -> int [2:18]
    |   |--IDENT -> var7 [2:22]
    |   `--ASSIGN -> = [2:27]
    |       `--EXPR -> EXPR [2:29]
    |           `--NUM_INT -> 6 [2:29]
    |--COMMA -> , [2:30]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:18]
    |   |--MODIFIERS -> MODIFIERS [2:18]
    |   |--TYPE -> TYPE [2:18]
    |   |   `--LITERAL_INT -> int [2:18]
    |   |--IDENT -> var8 [3:8]
    |   `--ASSIGN -> = [3:13]
    |       `--EXPR -> EXPR [3:15]
    |           `--NUM_INT -> 5 [3:15]
    |--SEMI -> ; [3:16]
    `--RCURLY -> } [4:0]
```
Is this the expected tree?

**Comment by nmancus1** [comment]
Ok, so I've been studying this for awhile, and the resulting Java code hoping to figure out what is happening here.  As far as I can tell, this: 

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L899-L902

is working as expected, creating comma separated variable definitions. I want to make sure I'm understanding what's supposed to be happening here:

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L825-829

1. Call `variableDefinitions` with modifiers
2. Assigning result to AST `v`
3. Checking if there is a `SEMI` to consume, and if so assigning it to rule `s6`.
4. Letting our AST `field` be AST `v`.
5. Appending the SEMI to the AST `v`.

It looks like the first "leaf" is consuming the SEMI because it is required by the variableDeclarator rule.

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L899-902

Is this all correct?

**Comment by nmancus1** [comment]
@rnveach is this what you had in mind, as far as the structure goes?  If this is OK, I will go ahead and create a DCL_LIST token and submit a PR.  I think that we can probably replace this https://github.com/checkstyle/checkstyle/blob/38a84fc4152f5a60db2161332c8df35ce7375d61/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L880-L887 with the new `VariableDeclaratorList`.

**Comment by rnveach** [comment]
> is this what you had in mind, as far as the structure goes?

I don't have a preference as long as it makes sense from a programmer who is trying to make a new check that uses the structure. If the structure doesn't follow a pattern, then sometimes you have to make weird overrides in the code when trying to use similar code on different AST types or the same AST type with a small difference added.

We try to replicate the JLS grammar so we can start with that and see what does/doesn't make sense.

**Comment by nmancus1** [comment]
The existing code always appends the SEMI to the first child.  I think that if we modify the `variableDeclarator` rule to grab the SEMI when parsing and attach it to the variable definition, we can avoid this issue.

https://github.com/checkstyle/checkstyle/commit/234a316caa2316b3d98774b4194383b8841a76e4

**Comment by rnveach** [comment]
Again I would have to see the actual tree being generated to give an opinion. If you have a solution, open a PR and tests will force you to showcase a tree for existing AST prints.

**Comment by nmancus1** [comment]
Is it ok to submit a PR that won't pass mvn clean verify?  I need some advice since my changes have made the OneStatementPerLineCheck unhappy, and I'm struggling to find a good solution.

**Comment by rnveach** [comment]
Yes, if you need help, starting PR is the best. I still want to see one case of printing new tree to see what we are now dealing with.
